proposals/p1928.md

```diff
@@ -0,0 +1,163 @@
+# Arrays
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/1928)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Problem
+
+Arrays are base component of almost of programming languages. In the current
+stage of Carbon development, arrays are missed.
+
+## Background
+
+An array is a set of scalar data placed in contiguous memory locations, all of
+the same type and type parameters. The scalar data that make up an array are the
+array elements which can be referenced by unique indexes.
+
+A nested array is an array of arrays which may be non-continuous in memory.
+
+A multidimensional array is an array with more than two dimensions which is
+continuous in memory.
+
+## Proposal
+
+This proposal introduces array internal representation, its declaration, and
+indexing.
+
+## Details
+
+### Internal representation
+
+For simplifying usage and raising safety of array usage, internal representation
+of arrays contains additional metadata:
+```cpp
+  struct {
+    int32_t dimensions;
+    int32_t * dimension_size;
+    void * data;
+  } Array;
+```
+Here, `dimensions` contains information about number of dimensions (necessary
+for multidimensional arrays), `dimension_size` is an array of size `dimensions`
+containing sizes for each dimensions, and `data` is continuous in memory place
+keeping scalar data.
+
+This representation is closer to `std::vector<T>` rather then `T[]`. Then,
+interoperability between C++ and carbon is limited.
+
+One-dimensional and nested arrays have `dimensions = 1`.
+
+### Declaration
```

Comment:
In accordance with the principle that [all APIs are library APIs](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/principles/library_apis_only.md), `[Type; Size]` will need to be an alias for some library class type, e.g. `Array(Type, Size)`. It would be good for this proposal to specify the name of that type.

Reply:
I took names from https://github.com/carbon-language/carbon-lang/discussions/1061